packages/name-controller/src/NameController.ts

```diff
@@ -0,0 +1,491 @@
+
+export type NameControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  NameControllerActions,
+  NameControllerEvents,
+  never,
+  never
+>;
+
+export type NameControllerOptions = {
+  getChainId: () => string;
+  messenger: NameControllerMessenger;
+  providers: NameProvider[];
+  state?: Partial<NameControllerState>;
+};
+
+export type UpdateProposedNamesRequest = {
+  value: string;
+  type: NameType;
+  sourceIds?: string[];
+};
+
+export type UpdateProposedNamesResult = {
+  results: Record<string, { proposedNames?: string[]; error?: unknown }>;
+};
+
+export type SetNameRequest = {
+  value: string;
+  type: NameType;
+  name: string;
+  sourceId?: string;
+};
+
+/**
+ * Controller for storing and deriving names for values such as Ethereum addresses.
+ */
+export class NameController extends BaseControllerV2<
+  typeof controllerName,
+  NameControllerState,
+  NameControllerMessenger
+> {
+  #getChainId: () => string;
+
+  #providers: NameProvider[];
+
+  /**
+   * Construct a Name controller.
+   *
+   * @param options - Controller options.
+   * @param options.getChainId - Callback that returns the chain ID of the current network.
+   * @param options.messenger - Restricted controller messenger for the name controller.
+   * @param options.providers - Array of name provider instances to propose names.
+   * @param options.state - Initial state to set on the controller.
+   */
+  constructor({
+    getChainId,
+    messenger,
+    providers,
+    state,
+  }: NameControllerOptions) {
+    super({
+      name: controllerName,
+      metadata: stateMetadata,
+      messenger,
+      state: { ...getDefaultState(), ...state },
+    });
+
+    this.#getChainId = getChainId;
+    this.#providers = providers;
+  }
+
+  /**
+   * Set the user specified name for a value.
+   *
+   * @param request - Request object.
+   * @param request.name - Name to set.
+   * @param request.sourceId - Optional ID of the source of the proposed name.
+   * @param request.type - Type of value to set the name for.
+   * @param request.value - Value to set the name for.
+   */
+  setName(request: SetNameRequest) {
+    this.#validateSetNameRequest(request);
+
+    const { value, type, name, sourceId } = request;
+
+    this.#updateEntry(value, type, { name, sourceId: sourceId ?? null });
+  }
+
+  /**
+   * Generate the proposed names for a value using the name providers and store them in the state.
+   *
+   * @param request - Request object.
+   * @param request.value - Value to update the proposed names for.
+   * @param request.type - Type of value to update the proposed names for.
+   * @param request.sourceIds - Optional array of source IDs to limit which sources are used by the providers. If not provided, all sources in all providers will be used.
+   * @returns The updated proposed names for the value.
+   */
+  async updateProposedNames(
+    request: UpdateProposedNamesRequest,
+  ): Promise<UpdateProposedNamesResult> {
```

Comment:
Sorry if my misunderstanding here.
Are we going to read proposed names from state of the `NameController` in client code? 
If yes, why do we need to return `UpdateProposedNamesResult`?
If no, how do we match proposed name - provider since `UpdateProposedNamesResult` doesn't return any info about provider?

Reply:
The frontends will indeed read the proposed names from the state, however if an alternate controller or any background logic invokes the method, I think the cleaner way to transfer the data would be the return value.

The `UpdateProposedNamesResult` keys the results by `sourceId` which indicates where the proposed names came from, such as `ens`, or `etherscan`.